micro_http/src/connection.rs

```diff
@@ -0,0 +1,714 @@
+                    // Update the state machine to expect a new request and push this request into
+                    // the `parsed_requests` queue.
+                    self.state = ConnectionState::WaitingForRequestLine;
+                    self.body_bytes_to_be_read = 0;
+                    self.parsed_requests
+                        .push_back(self.pending_request.take().unwrap());
+                }
+            };
+        }
+    }
+
+    fn read_bytes(&mut self) -> Result<usize, ConnectionError> {
+        // Append new bytes to what we already have in the buffer.
+        let bytes_read = self
+            .stream
+            .read(&mut self.buffer[self.read_cursor..])
+            .map_err(|_| ConnectionError::StreamError)?;
+
+        // If the read returned 0 then the client has closed the connection.
+        if bytes_read == 0 {
+            return Err(ConnectionError::ClosedConnection);
+        }
+
+        // `end_cursor` will be at the end of what we have just appended.
+        Ok(bytes_read + self.read_cursor)
+    }
+
+    fn read_request_line(
+        &mut self,
+        end_cursor: usize,
+        line_start_index: &mut usize,
+    ) -> Result<bool, ConnectionError> {
+        match find(&self.buffer[*line_start_index..end_cursor], &[CR, LF]) {
+            Some(line_end_index) => {
+                let line = &self.buffer[*line_start_index..(*line_start_index + line_end_index)];
+
+                *line_start_index = *line_start_index + line_end_index + CRLF_LEN;
+                let request_line =
+                    RequestLine::try_from(line).map_err(ConnectionError::ParseError)?;
+
+                // Form the request with a valid request line, which is the bare minimum
+                // for a valid request.
+                self.pending_request = Some(Request {
+                    request_line,
+                    headers: Headers::default(),
+                    body: None,
+                });
+                self.state = ConnectionState::WaitingForHeaders;
+                Ok(true)
+            }
+            None => {
+                // The request line is longer than BUFFER_SIZE bytes, so the request is invalid.
+                if end_cursor == BUFFER_SIZE && *line_start_index == 0 {
+                    return Err(ConnectionError::ParseError(RequestError::InvalidRequest));
+                } else {
+                    // Move the incomplete request line to the beginning of the buffer and wait
+                    // for the next `try_read` call to complete it.
+                    self.shift_buffer_left(*line_start_index, end_cursor);
+                }
+                Ok(false)
+            }
+        }
+    }
+
+    fn read_headers(
+        &mut self,
+        end_cursor: usize,
+        line_start_index: &mut usize,
+    ) -> Result<bool, ConnectionError> {
+        match find(&self.buffer[*line_start_index..end_cursor], &[CR, LF]) {
+            // We have found the end of the headers.
```

Comment:
Explain the logic here. Why is it the end of the headers? Most people looking at this code won't be familiar with the exact HTTP definition.

Reply:
This comment was not resolved.